apps/rush-lib/src/cli/utilities/pnpm/PnpmLinkManager.ts

```diff
@@ -0,0 +1,210 @@
+// See LICENSE in the project root for license information.
+
+import * as fsx from 'fs-extra';
+import * as os from 'os';
+import * as path from 'path';
+import * as semver from 'semver';
+import encodeRegistry = require('encode-registry');
+
+import { JsonFile } from '@microsoft/node-core-library';
+
+import {
+  BaseLinkManager,
+  SymlinkKind
+} from '../base/BaseLinkManager';
+import Utilities from '../../../utilities/Utilities';
+import { BasePackage } from '../base/BasePackage';
+import { RushConstants } from '../../../RushConstants';
+import { IRushLinkJson } from '../../../data/RushConfiguration';
+import RushConfigurationProject from '../../../data/RushConfigurationProject';
+
+export class PnpmLinkManager extends BaseLinkManager {
+  protected _linkProjects(): Promise<void> {
+    return new Promise<void>((resolve: () => void, reject: (reason: Error) => void): void => {
+      // go ahead and find out the registry we are using, we need this to locate the pnpm store folder
+
+      // EXAMPLE: https://onedrive.pkgs.visualstudio.com/_packaging/odsp-npm/npm/registry
+      const registryUrl: string = Utilities.executeCommandAndCaptureOutput(
+        this._rushConfiguration.packageManagerToolFilename,
+        ['config', 'get', 'registry'],
+        this._rushConfiguration.commonTempFolder,
+        process.env);
+
+      // EXAMPLE: onedrive.pkgs.visualstudio.com
+      const encodedRegistry: string = encodeRegistry(registryUrl);
+
+      // EXAMPLE: c:/src/foo/common/temp/node_modules/.onedrive.pkgs.visualstudio.com
+      const registryPath: string = path.join(
+        this._rushConfiguration.commonTempFolder,
+        RushConstants.nodeModulesFolderName,
+        `.${encodedRegistry}`
+      );
+
+      try {
+        const rushLinkJson: IRushLinkJson = { localLinks: {} };
+
+        for (const rushProject of this._rushConfiguration.projects) {
+          console.log(os.EOL + 'LINKING: ' + rushProject.packageName);
+          this._linkProject(registryPath, rushProject, rushLinkJson);
+        }
+
+        console.log(`Writing "${this._rushConfiguration.rushLinkJsonFilename}"`);
+        JsonFile.save(rushLinkJson, this._rushConfiguration.rushLinkJsonFilename);
+
+        resolve();
+      } catch (error) {
+        reject(error);
+      }
+    });
+  }
+
+  /**
+   * This is called once for each local project from Rush.json.
+   * @param project             The local project that we will create symlinks for
+   * @param rushLinkJson        The common/temp/rush-link.json output file
+   */
+  private _linkProject(
+    pnpmStorePath: string,
+    project: RushConfigurationProject,
+    rushLinkJson: IRushLinkJson): void {
+
+    // PNPM creates a directed acyclic graph, rather than a tree.
+    // Thus, imagine "sp-http" and "sp-core-library" both depend on "request"@1.2.3,
+    // and "request"@1.2.3 depends on "http@*", then both "sp-http" and "sp-core-library"
```

Comment:
>http [](start = 39, length = 4)

Can you choose examples that are a little more general and with names that aren't confusing or hard to say (sp-http vs http)